avalanche/spec.md

```diff
@@ -24,7 +24,12 @@ If a node receives any avalanche message (from below) from a peer not signaling
 | ?          | signature        | []byte    | A signature covering the remote peer's version nonce |
 
 The `avapubkey` message must be sent to any remote peer signaling `NODE_AVALANCHE` after the receipt of the `version` message. The message consists of
-a compressed secp256k1 public key and a signature covering the nonce found in the remote peer's version message. The Bitcoin protocol does not
+a compressed secp256k1 public key and a signature covering the nonce found in the remote peer's version message.
+
+> What is the purpose of the pubkey?  Sybil resistance?  We need to know how this works.
+> Sybil attack vectors are: compromise local DNS (seeder), compromise WAN add/drop mux to isolate and MITM your data via CALEA, P2P -- outlast other peers and only recommend your sybils.
+
```

Comment:
So as you mention later on the sybil resistence mechanism isn't defined yet. I had in mind something like the following when creating this implementation...

- Miners put a pubkey in the coinbase
- Node tracks the pubkeys of the last n blocks. 
- Node's authenticate themselves using their pubkey... for miners this would match the pubkey found in their coinbases. 
- the `getPeerToQuery()` function would select a random peer weighted by the number of times their pubkey appears in the last n blocks. 
- the privkey is used to sign the `avaresponse` so the node knows the response has not been tampered with.

Reply:
Like deadlnix has pointed out, different membership sets at different tips is going to be a PITA.

Reply:
It's possible we could put them into groups based on "epochs" instead of changing every single block.

Reply:
So the proposal is that you can only vote if you've mined a block within the last n blocks?  How will a node know how to contact a voting miner?  What are the effects on avalanche is one entity gets to make (say) 30-40% of the votes?

WRT signing avaresponse with the privkey, it looked to me like this was the privkey of a pubkey provided earlier in the protocol?  That strategy won't work because if you can MITM the protocol, you could inject your own pubkey.  The signing privkey needs to be the one in the coinbase.

This still leaves open the DOS attack vector which we'll need to think about mitigation strategies.